File: Segundo_Curso/Segundo_Cuatri/PDOO/Estudio/ruby/Irrgarten/LabyrinthCharacter.rb
module Irrgarten
  class LabyrinthCharacter
    @@INVALID_POS = -1    # Se hereda

    # Inicializador de LabyrinthCharacter, no tiene numero pues los monstruos no se enumeran.
    # Si no hacemos uso de setStats que provoca que se ejecute el código de este archivo, en un hijo se ejecutará el initialize del hijo por la ligadura dinamica.
    def initialize(name,intelligence,strength,health)
      setStats(name,intelligence,strength,health)
      @row = @@INVALID_POS
      @col = @@INVALID_POS
    end

    #Creado porque si se sobrecarga initialize no se puede usar de forma segura
    #De esta forma podemos en el cloner delegar aqui. Además evita repeticion de codigo
    def setStats(name, intelligence,strength,health)
      @name = name
      @intelligence = intelligence.to_f
      @strength = strength.to_f
      @health = health
    end
    private :setStats

    attr_reader :name
    attr_reader :row
    attr_reader :col

    # Constructor de copia de un Player
    # other -> Personaje del que copiar
    def cloner(other)
      setStats(other.name,other.intelligence, other.strength,other.health)
      @row = other.row
      @col = other.col
    end

    # Comprueba si un jugador esta muerto
    # return -> True -> muerto
    #           False -> vivo
    def dead
      @health<=0
    end

    protected
      # getter intelligence
      def intelligence
        @intelligence
      end

      # getter strength
      def strength
        @strength
      end

      # getter health
      def health
        @health
      end

      # setter health
      # health -> nuevo valor de vida
      def health=(health)
        @health = health
      end

      # Daña a un LabyrinthCharacter
      def gotWounded
        @health -= 1
      end
    public

      # setter row y col
      # row -> fila nueva
      # col -> columna nueva
      def setPos(row,col)
        if (row >= 0) && (col >= 0) then
          @row = row
          @col = col
        end
      end

      # Concatena la informacion de un LabyrinthCharacter
      # return -> cadena con la informacion
      def to_s
        "[#{@name}, #{@health} HP, #{@intelligence} IP, #{@strength} SP,(#{@row},#{@col})]\n"
      end

  end#class
end#module

File: Segundo_Curso/Segundo_Cuatri/PDOO/Estudio/java/src/irrgarten/LabyrinthCharacter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author el_dramas
 */
public abstract class LabyrinthCharacter {
    
    static private final int INVALID_POS = -1;
    
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    
    /**
     * Constructor de LabyrinthCharacter.
     * @param name
     * @param intelligence
     * @param strength
     * @param health 
     */
    public LabyrinthCharacter(String name,float intelligence,float strength, float health){
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = health;
        
        this.row = INVALID_POS;
        this.col = INVALID_POS; 
    }
    
    /**
     * Constructor de copia de un LabyrinthCharacter
     * @param other Instancia de LabyrinthCharacter
     */
    public LabyrinthCharacter(LabyrinthCharacter other){
        this (other.name,other.getIntelligence(),other.getStrength(),other.getHealth());    //Llama al constructor de LabyrinthCharacter
        this.row = other.row;
        this.col = other.col;
        //this.setPos(other.getRow(), other.getCol()); Da warning de overridable de setPos, es decir, si se sobreescribe no se ejecuta este metodo sino el de la ligadura dinamica

    }
    
    /**
     * Comprueba si el LabyrinthCharacter esta muerto.
     * @return true -> health <= 0;
     *         false -> health > 0;
     */
    boolean dead(){
        return this.health <= 0;
    }
    
    /**
     * Devuelve la fila.
     * @return Fila del LabyrinthCharacter.
     */
    public int getRow(){
        return this.row;
    }
    
    /**
     * Devuelve la columna.
     * @return Columna del LabyrinthCharacter.
     */
    public int getCol(){
        return this.col;
    }
    
    /**
     * Devuelve la inteligencia.
     * @return Inteligencia del LabyrinthCharacter.
     */
    protected float getIntelligence(){
        return this.intelligence;
    }
    
    /**
     * Devuelve la fuerza.
     * @return Fuerza del LabyrinthCharacter.
     */
    protected float getStrength(){
        return this.strength;
    }
    
    /**
     * Devuelve la salud.
     * @return Salud del LabyrinthCharacter.
     */
    protected float getHealth(){
        return this.health;
    }
    
    /**
     * Almacena la salud.
     * @param health salud del LabyrinthCharacter
     */
    protected void setHealth(float health){
        this.health = health;
    }
    
    /**
     * Impone la posicion del LabyrinthCharacter.
     * @param row Fila del LabyrinthCharacter.
     * @param col Columna del LabyrinthCharacter.
     */
    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }
    
    /**
     * Concatena la informacion de una instancia.
     * @return Cadena con la informacion.
     */
    @Override
    public String toString(){
        return "[" + name + ", " + health +" HP," + strength + " SP, " +  intelligence + " IP, " + "(" + row + ", " +col + ")]";
    }
    
    /**
     * Maneja el daño de un LabyrinthCharacter.
     */
    protected void gotWounded(){
        this.health--;
    }
    
    /**
     * Maneja el ataque de un LabyrinthCharacter.
     * @return Valor de ataque.
     */
    public abstract float attack();
    
    /**
     * Maneja la defensa de un LabyrinthCharacter.
     * @param attack Ataque recibido
     * @return true -> perder por el golpe
     */
    public abstract boolean defend(float attack);

}
